File: src/server/scanner/detectors.ts
import { existsSync, readFileSync } from 'fs';
import { join } from 'path';
import type { ProjectType, DeployTarget } from '../../shared/types.js';

interface DetectionResult {
  type: ProjectType;
  techStack: string[];
  devCommand: string | null;
  devPort: number | null;
  deployTarget: DeployTarget;
  description: string | null;
}

function fileExists(dir: string, file: string): boolean {
  return existsSync(join(dir, file));
}

function dirExists(dir: string, sub: string): boolean {
  try {
    const stat = Bun.file(join(dir, sub));
    // Bun.file doesn't check dirs — use fs
    return existsSync(join(dir, sub));
  } catch {
    return false;
  }
}

function readPkg(dir: string): Record<string, unknown> | null {
  const pkgPath = join(dir, 'package.json');
  if (!existsSync(pkgPath)) return null;
  try {
    return JSON.parse(readFileSync(pkgPath, 'utf-8'));
  } catch {
    return null;
  }
}

function hasDep(pkg: Record<string, unknown>, dep: string): boolean {
  const deps = { ...(pkg.dependencies as Record<string, string> || {}), ...(pkg.devDependencies as Record<string, string> || {}) };
  return dep in deps;
}

function extractPort(dir: string, pkg: Record<string, unknown> | null): number | null {
  // Check vite config for port
  for (const cfg of ['vite.config.ts', 'vite.config.js']) {
    const p = join(dir, cfg);
    if (existsSync(p)) {
      try {
        const content = readFileSync(p, 'utf-8');
        const match = content.match(/port\s*:\s*(\d+)/);
        if (match) return parseInt(match[1]);
      } catch { /* skip */ }
    }
  }

  // Check next.config for port in dev script
  if (pkg) {
    const scripts = pkg.scripts as Record<string, string> || {};
    const devCmd = scripts.dev || '';
    const portMatch = devCmd.match(/-p\s+(\d+)|--port\s+(\d+)|PORT=(\d+)/);
    if (portMatch) return parseInt(portMatch[1] || portMatch[2] || portMatch[3]);
  }

  // Default ports by type
  return null;
}

export function detectProject(dir: string): DetectionResult {
  const pkg = readPkg(dir);
  const techStack: string[] = [];
  let type: ProjectType = 'unknown';
  let devCommand: string | null = null;
  let devPort: number | null = null;
  let deployTarget: DeployTarget = 'none';
  let description: string | null = null;

  if (pkg) {
    description = (pkg.description as string) || null;
    const scripts = pkg.scripts as Record<string, string> || {};
    devCommand = scripts.dev || scripts.start || null;

    // Detect tech stack from deps
    if (hasDep(pkg, 'react')) techStack.push('react');
    if (hasDep(pkg, 'typescript') || fileExists(dir, 'tsconfig.json')) techStack.push('typescript');
    if (hasDep(pkg, 'tailwindcss') || hasDep(pkg, '@tailwindcss/vite')) techStack.push('tailwind');
    if (hasDep(pkg, 'vite')) techStack.push('vite');
    if (hasDep(pkg, 'next')) techStack.push('next');
    if (hasDep(pkg, 'hono')) techStack.push('hono');
    if (hasDep(pkg, 'supabase') || hasDep(pkg, '@supabase/supabase-js')) techStack.push('supabase');
    if (hasDep(pkg, 'better-sqlite3') || hasDep(pkg, 'sqlite3')) techStack.push('sqlite');
    if (hasDep(pkg, 'playwright') || hasDep(pkg, '@playwright/test')) techStack.push('playwright');
    if (hasDep(pkg, 'framer-motion') || hasDep(pkg, 'motion')) techStack.push('framer-motion');
    if (hasDep(pkg, 'stripe') || hasDep(pkg, '@stripe/stripe-js')) techStack.push('stripe');
    if (hasDep(pkg, '@anthropic-ai/sdk')) techStack.push('anthropic');
    if (hasDep(pkg, 'vitest')) techStack.push('vitest');
    if (hasDep(pkg, '@dnd-kit/core')) techStack.push('dnd-kit');
    if (hasDep(pkg, 'three') || hasDep(pkg, '@react-three/fiber')) techStack.push('three.js');
    if (hasDep(pkg, 'vue')) techStack.push('vue');
    if (hasDep(pkg, 'svelte')) techStack.push('svelte');
    if (hasDep(pkg, 'd3') || hasDep(pkg, 'd3-selection')) techStack.push('d3');
  }

  // Detect project type (priority order)
  if (dirExists(dir, 'CSXS') || fileExists(dir, '.debug')) {
    type = 'cep-plugin';
    // Detect shared libraries from config
    try {
      const { getUserConfig } = require('../userConfig.js');
      const uc = getUserConfig();
      for (const lib of uc.sharedLibraries) {
        if (dirExists(dir, lib.subdir)) techStack.push(lib.name.toLowerCase());
      }
    } catch { /* config not ready */ }
  } else if (fileExists(dir, 'next.config.ts') || fileExists(dir, 'next.config.js') || fileExists(dir, 'next.config.mjs')) {
    type = 'nextjs';
  } else if (fileExists(dir, 'framer.json')) {
    type = 'framer-plugin';
  } else if ((fileExists(dir, 'vite.config.ts') || fileExists(dir, 'vite.config.js')) && pkg && hasDep(pkg, 'react')) {
    type = 'vite-react';
  } else if (fileExists(dir, 'wrangler.toml')) {
    type = 'cloudflare-worker';
    deployTarget = 'cloudflare';
  } else if (pkg && hasDep(pkg, 'hono') && !hasDep(pkg, 'react')) {
    type = 'hono-server';
  } else if (fileExists(dir, 'Package.swift')) {
    type = 'swift-app';
    techStack.push('swift');
  } else if (pkg && (pkg.main || pkg.exports)) {
    type = 'node-package';
  } else if (fileExists(dir, 'index.html')) {
    type = 'static-site';
  }

  // Detect deploy target
  if (fileExists(dir, 'vercel.json') || dirExists(dir, '.vercel')) {
    deployTarget = 'vercel';
  } else if (fileExists(dir, 'wrangler.toml')) {
    deployTarget = 'cloudflare';
  } else if (fileExists(dir, 'netlify.toml') || dirExists(dir, '.netlify')) {
    deployTarget = 'netlify';
  } else if (fileExists(dir, 'railway.json') || fileExists(dir, 'railway.toml')) {
    deployTarget = 'railway';
  } else if (fileExists(dir, 'fly.toml')) {
    deployTarget = 'flyio';
  }

  // Docker detection
  if (fileExists(dir, 'docker-compose.yml') || fileExists(dir, 'docker-compose.yaml') || fileExists(dir, 'compose.yml') || fileExists(dir, 'compose.yaml')) {
    techStack.push('docker-compose');
  }
  if (fileExists(dir, 'Dockerfile')) {
    techStack.push('docker');
  }

  devPort = extractPort(dir, pkg);

  return { type, techStack, devCommand, devPort, deployTarget, description };
}

File: Sources/DevDock/Server/Scanner/Detectors.swift
import Foundation

/// Project type + tech stack + deploy target detection.
///
/// Ported from `src/server/scanner/detectors.ts`. Detection priority order
/// matches the TS version exactly so existing projects get the same type
/// when re-scanned.
enum Detectors {

    struct Result {
        var type: String = "unknown"
        var techStack: [String] = []
        var devCommand: String? = nil
        var devPort: Int? = nil
        var deployTarget: String = "none"
        var description: String? = nil
    }

    static func detect(dir: String) -> Result {
        var result = Result()
        let pkg = readPkg(dir: dir)

        if let pkg = pkg {
            result.description = pkg["description"] as? String

            // Pick package manager by lockfile
            let hasBun = fileExists(dir, "bun.lock") || fileExists(dir, "bun.lockb")
            let hasPnpm = fileExists(dir, "pnpm-lock.yaml")
            let hasYarn = fileExists(dir, "yarn.lock")
            let pm: String
            if hasBun { pm = "bun" }
            else if hasPnpm { pm = "pnpm" }
            else if hasYarn { pm = "yarn" }
            else { pm = "npm" }

            if let scripts = pkg["scripts"] as? [String: String] {
                if scripts["dev"] != nil {
                    result.devCommand = "\(pm) run dev"
                } else if scripts["start"] != nil {
                    result.devCommand = "\(pm) run start"
                }
            }

            // Tech stack from deps
            let deps = collectDeps(pkg)
            if deps.contains("react") { result.techStack.append("react") }
            if deps.contains("typescript") || fileExists(dir, "tsconfig.json") { result.techStack.append("typescript") }
            if deps.contains("tailwindcss") || deps.contains("@tailwindcss/vite") { result.techStack.append("tailwind") }
            if deps.contains("vite") { result.techStack.append("vite") }
            if deps.contains("next") { result.techStack.append("next") }
            if deps.contains("hono") { result.techStack.append("hono") }
            if deps.contains("supabase") || deps.contains("@supabase/supabase-js") { result.techStack.append("supabase") }
            if deps.contains("better-sqlite3") || deps.contains("sqlite3") { result.techStack.append("sqlite") }
            if deps.contains("playwright") || deps.contains("@playwright/test") { result.techStack.append("playwright") }
            if deps.contains("framer-motion") || deps.contains("motion") { result.techStack.append("framer-motion") }
            if deps.contains("stripe") || deps.contains("@stripe/stripe-js") { result.techStack.append("stripe") }
            if deps.contains("@anthropic-ai/sdk") { result.techStack.append("anthropic") }
            if deps.contains("vitest") { result.techStack.append("vitest") }
            if deps.contains("@dnd-kit/core") { result.techStack.append("dnd-kit") }
            if deps.contains("three") || deps.contains("@react-three/fiber") { result.techStack.append("three.js") }
            if deps.contains("vue") { result.techStack.append("vue") }
            if deps.contains("svelte") { result.techStack.append("svelte") }
            if deps.contains("d3") || deps.contains("d3-selection") { result.techStack.append("d3") }
        }

        // Project type (priority order — first match wins)
        if dirExists(dir, "CSXS") || fileExists(dir, ".debug") {
            result.type = "cep-plugin"
            // Shared library detection
            let uc = ConfigStore.shared.load()
            for lib in uc.sharedLibraries where dirExists(dir, lib.subdir) {
                result.techStack.append(lib.name.lowercased())
            }
        } else if fileExists(dir, "next.config.ts") || fileExists(dir, "next.config.js") || fileExists(dir, "next.config.mjs") {
            result.type = "nextjs"
        } else if fileExists(dir, "framer.json") {
            result.type = "framer-plugin"
        } else if (fileExists(dir, "vite.config.ts") || fileExists(dir, "vite.config.js")) && pkg != nil && hasDep(pkg!, "react") {
            result.type = "vite-react"
        } else if fileExists(dir, "wrangler.toml") {
            result.type = "cloudflare-worker"
            result.deployTarget = "cloudflare"
        } else if let pkg, hasDep(pkg, "hono"), !hasDep(pkg, "react") {
            result.type = "hono-server"
        } else if fileExists(dir, "Package.swift") {
            result.type = "swift-app"
            result.techStack.append("swift")
        } else if let pkg, (pkg["main"] != nil || pkg["exports"] != nil) {
            result.type = "node-package"
        } else if fileExists(dir, "index.html") {
            result.type = "static-site"
        }

        // Deploy target (may override cloudflare set above if another present)
        if fileExists(dir, "vercel.json") || dirExists(dir, ".vercel") {
            result.deployTarget = "vercel"
        } else if fileExists(dir, "wrangler.toml") {
            result.deployTarget = "cloudflare"
        } else if fileExists(dir, "netlify.toml") || dirExists(dir, ".netlify") {
            result.deployTarget = "netlify"
        } else if fileExists(dir, "railway.json") || fileExists(dir, "railway.toml") {
            result.deployTarget = "railway"
        } else if fileExists(dir, "fly.toml") {
            result.deployTarget = "flyio"
        }

        // Docker detection
        if fileExists(dir, "docker-compose.yml") || fileExists(dir, "docker-compose.yaml")
            || fileExists(dir, "compose.yml") || fileExists(dir, "compose.yaml") {
            result.techStack.append("docker-compose")
        }
        if fileExists(dir, "Dockerfile") {
            result.techStack.append("docker")
        }

        result.devPort = extractPort(dir: dir, pkg: pkg)

        return result
    }

    // MARK: - Port extraction

    private static func extractPort(dir: String, pkg: [String: Any]?) -> Int? {
        // vite.config.{ts,js} — look for `port: NNNN`
        for cfg in ["vite.config.ts", "vite.config.js"] {
            let path = (dir as NSString).appendingPathComponent(cfg)
            if let content = try? String(contentsOfFile: path, encoding: .utf8),
               let match = content.range(of: #"port\s*:\s*(\d+)"#, options: .regularExpression) {
                let substring = String(content[match])
                if let num = substring.components(separatedBy: CharacterSet.decimalDigits.inverted)
                    .joined(separator: "")
                    .nonEmpty(),
                   let n = Int(num) {
                    return n
                }
            }
        }

        // package.json dev script — -p NNNN | --port NNNN | PORT=NNNN
        if let pkg, let scripts = pkg["scripts"] as? [String: String], let dev = scripts["dev"] {
            let pattern = #"-p\s+(\d+)|--port\s+(\d+)|PORT=(\d+)"#
            if let re = try? NSRegularExpression(pattern: pattern),
               let match = re.firstMatch(in: dev, range: NSRange(dev.startIndex..., in: dev)) {
                for groupIdx in 1...3 {
                    let range = match.range(at: groupIdx)
                    if range.location != NSNotFound, let r = Range(range, in: dev) {
                        return Int(dev[r])
                    }
                }
            }
        }

        return nil
    }

    // MARK: - Helpers

    private static func fileExists(_ dir: String, _ name: String) -> Bool {
        FileManager.default.fileExists(atPath: (dir as NSString).appendingPathComponent(name))
    }

    private static func dirExists(_ dir: String, _ sub: String) -> Bool {
        var isDir: ObjCBool = false
        let path = (dir as NSString).appendingPathComponent(sub)
        return FileManager.default.fileExists(atPath: path, isDirectory: &isDir) && isDir.boolValue
    }

    private static func readPkg(dir: String) -> [String: Any]? {
        let path = (dir as NSString).appendingPathComponent("package.json")
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return obj
    }

    private static func collectDeps(_ pkg: [String: Any]) -> Set<String> {
        var out = Set<String>()
        if let deps = pkg["dependencies"] as? [String: Any] {
            for key in deps.keys { out.insert(key) }
        }
        if let deps = pkg["devDependencies"] as? [String: Any] {
            for key in deps.keys { out.insert(key) }
        }
        return out
    }

    private static func hasDep(_ pkg: [String: Any], _ name: String) -> Bool {
        if let deps = pkg["dependencies"] as? [String: Any], deps[name] != nil { return true }
        if let deps = pkg["devDependencies"] as? [String: Any], deps[name] != nil { return true }
        return false
    }
}

private extension String {
    func nonEmpty() -> String? { isEmpty ? nil : self }
}
